LFS-1223: Create a utility program for backing up and restoring the tag of a running Docker image

IntegralProgrammer opened the pull request:
This PR implements LFS-1223 by providing the `backup_image_tag.sh` and `restore_image_tag.sh` utility programs. These utility programs will be used when implementing LFS-1214 so that if updating a Docker image causes something to break, we can easily roll back to the previously working version of said Docker image.

To test:
1. Pull the latest Docker httpd image (`docker pull httpd`)
2. Tag this `httpd:latest` image as `nginx:latest` (`docker image tag httpd:latest nginx:latest`)
3. Start the mock `nginx:latest` image which is in reality `httpd:latest` (`docker run --rm -p 8080:80 -d nginx:latest`). Visiting http://localhost:8080 should load an Apache Httpd page
4. Stop this container
5. Backup the `nginx:latest` image tag (`./Utilities/ContainerManagement/backup_image_tag.sh nginx:latest ~/nginx.tagbackup`)
6. Pull the real `nginx:latest` (`docker pull nginx`)
7. Start the nginx image with the same command as before (`docker run --rm -p 8080:80 -d nginx:latest`). Visiting http://localhost:8080 should load an Nginx page.
8. Stop this container
9. Restore the backed-up `nginx:latest` tag (`./Utilities/ContainerManagement/restore_image_tag.sh nginx:latest ~/nginx.tagbackup`)
10. Start a `nginx:latest` container again (`docker run --rm -p 8080:80 -d nginx:latest`). Visiting http://localhost:8080 should load an Apache Httpd page as in step 9 we have restored the `nginx:latest` tag to its backed-up value which is an Apache httpd Docker image.

IntegralProgrammer committed:
LFS-1223: Create a utility program for backing up and restoring the tag of a running Docker image

sdumitriu commented:
Running `./Utilities/ContainerManagement/backup_image_tag.sh lfs/lfs:0.8.0 bk.txt`:

> Template parsing error: template: :1:2: executing "" at <index .RepoDigests 0>: error calling index: reflect: slice index out of range

IntegralProgrammer commented:
> Running `./Utilities/ContainerManagement/backup_image_tag.sh lfs/lfs:0.8.0 bk.txt`:
> 
> > Template parsing error: template: :1:2: executing "" at <index .RepoDigests 0>: error calling index: reflect: slice index out of range

I'm not sure why you're getting that error. When you run `docker image inspect lfs/lfs:0.8.0`, what is the output?

sdumitriu commented:
> I'm not sure why you're getting that error. When you run `docker image inspect lfs/lfs:0.8.0`, what is the output?

```
[
    {
        "Id": "sha256:5793413f3fa4e3fdf96b2c0bdb39f70b36526e9fd0a933daa34c449d5bf4cfe9",
        "RepoTags": [
            "lfs/lfs:0.8.0"
        ],
        "RepoDigests": [],
        "Parent": "sha256:65fe72d5df4cb3726dfca7358bbc9f089ddb8a1125d6439efdc06afec7548e3c",
        "Comment": "",
        "Created": "2021-03-04T17:25:26.229963919Z",
        "Container": "600de88b782219821c0e3b9ebd20607081dc69a25eeb46177519ca70452047ba",
        "ContainerConfig": {
            "Hostname": "600de88b7822",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "5005/tcp": {},
                "8080/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/lib/jvm/java-1.8-openjdk/jre/bin:/usr/lib/jvm/java-1.8-openjdk/bin",
                "LANG=C.UTF-8",
                "JAVA_HOME=/usr/lib/jvm/java-1.8-openjdk/jre",
                "JAVA_VERSION=8u212",
                "JAVA_ALPINE_VERSION=8.212.04-r0",
                "DEBUG=",
                "DEV="
            ],
            "Cmd": [
                "/bin/sh",
                "-c",
                "#(nop) ",
                "ENTRYPOINT [\"/bin/sh\" \"-c\" \"/docker_entry.sh lfs 0.8.0\"]"
            ],
            "Image": "sha256:65fe72d5df4cb3726dfca7358bbc9f089ddb8a1125d6439efdc06afec7548e3c",
            "Volumes": {
                "/opt/lfs/sling": {}
            },
            "WorkingDir": "/opt/lfs",
            "Entrypoint": [
                "/bin/sh",
                "-c",
                "/docker_entry.sh lfs 0.8.0"
            ],
            "OnBuild": null,
            "Labels": {}
        },
        "DockerVersion": "20.10.4",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "5005/tcp": {},
                "8080/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/lib/jvm/java-1.8-openjdk/jre/bin:/usr/lib/jvm/java-1.8-openjdk/bin",
                "LANG=C.UTF-8",
                "JAVA_HOME=/usr/lib/jvm/java-1.8-openjdk/jre",
                "JAVA_VERSION=8u212",
                "JAVA_ALPINE_VERSION=8.212.04-r0",
                "DEBUG=",
                "DEV="
            ],
            "Cmd": null,
            "Image": "sha256:65fe72d5df4cb3726dfca7358bbc9f089ddb8a1125d6439efdc06afec7548e3c",
            "Volumes": {
                "/opt/lfs/sling": {}
            },
            "WorkingDir": "/opt/lfs/",
            "Entrypoint": [
                "/bin/sh",
                "-c",
                "/docker_entry.sh lfs 0.8.0"
            ],
            "OnBuild": null,
            "Labels": null
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 367404909,
        "VirtualSize": 367404909,
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/f6ba2c050d9b7ca9441574e605da69eebf3d0c9d6e08cb608a98186cf2a39bb1/diff:/var/lib/docker/overlay2/c307722a037b9cac13dffd04be423a3129298d05ddf1ac34b6ad83c7a7f94813/diff:/var/lib/docker/overlay2/dc8c0b853a54709c8e088c16482a9396aed500b4ba0ca5db0692765abcb7b3c7/diff:/var/lib/docker/overlay2/91975d91f549ef6e8574185b181b63e74a5312e8307195e25e547bf3a6433de2/diff:/var/lib/docker/overlay2/a51f60f2ce62cdb33f885e81378cc791811b6cbe7399a0b0679fb08e04beff25/diff:/var/lib/docker/overlay2/0699fa5f2d4f6737dcd084c521fece83cace75bf061e9c1f0b1753cc85297a8e/diff:/var/lib/docker/overlay2/7134343b777c39aef9253969cab06ab4d6513d76bb4feebcbed52aa4747ec7e6/diff:/var/lib/docker/overlay2/a710269e9ec1d40a919279665e0e65e081bdf24468827259130ac0a3e2ac1e18/diff",
                "MergedDir": "/var/lib/docker/overlay2/61e87f18a35c1a2dc90e7d851a1fe895913a8a9bb8b380b168fe87a1dfcbfc99/merged",
                "UpperDir": "/var/lib/docker/overlay2/61e87f18a35c1a2dc90e7d851a1fe895913a8a9bb8b380b168fe87a1dfcbfc99/diff",
                "WorkDir": "/var/lib/docker/overlay2/61e87f18a35c1a2dc90e7d851a1fe895913a8a9bb8b380b168fe87a1dfcbfc99/work"
            },
            "Name": "overlay2"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:f1b5933fe4b5f49bbe8258745cf396afe07e625bdab3168e364daf7c956b6b81",
                "sha256:9b9b7f3d56a01e3d9076874990c62e7a516cc4032f784f421574d06b18ef9aa4",
                "sha256:edd61588d12669e2d71a0de2aab96add3304bf565730e1e6144ec3c3fac339e4",
                "sha256:441f850edfc216c99d1e76ec1f1fc26503cf033695c9b33afae52066a1baa69d",
                "sha256:2c80f90fadf3c9c9d520fadd3c8abafba56233f73f136698de9e18863983a9b0",
                "sha256:d04ed13beb9dcb96de6852dd225a78ca7c835fd2d165d2e27252d5d833ed9108",
                "sha256:c007cf39611dfc06e90db93f552cb9c4c65d3acccada6b3ae3cf0c58aa784d86",
                "sha256:42dfc4bce92ed50c888b759c34bdfe0c6ad4ab93a86517a334826541259f7c7f",
                "sha256:46379271168be6c2feb9b314b9a19d76175bb0915ce90d527e2db83d559d8602"
            ]
        },
        "Metadata": {
            "LastTagTime": "2021-03-04T12:25:27.087381332-05:00"
        }
    }
]
```

IntegralProgrammer commented:
From what I understand, the reason why you are seeing an empty `RepoDigests` property is because the image that you are inspecting (`lfs/lfs:0.8.0`) was built locally and not pulled from a repository. Do you also get this error with images pulled from Docker Hub?

Given that these scripts are only to be used with our dependency Docker containers pulled from Docker Hub (eg. `mongo`, `ubuntu`) I don't see this as a problem. If you're able to use these scripts with images from Docker Hub, should I simply modify the `backup_image_tag.sh` script so that, if `RepoDigests` is empty, then an `Error: Please ensure that the source image was obtained from a repository and not locally` message is printed to STDERR and the command exits with a non-zero exit status?

sdumitriu commented:
> Do you also get this error with images pulled from Docker Hub?

No.

> Given that these scripts are only to be used with our dependency Docker containers pulled from Docker Hub (eg. `mongo`, `ubuntu`) I don't see this as a problem. If you're able to use these scripts with images from Docker Hub, should I simply modify the `backup_image_tag.sh` script so that, if `RepoDigests` is empty, then an `Error: Please ensure that the source image was obtained from a repository and not locally` message is printed to STDERR and the command exits with a non-zero exit status?

Sounds good.

IntegralProgrammer committed:
LFS-1223: Create a utility program for backing up and restoring the tag of a running Docker image

If the user tries to use backup_image_tag.sh on an image that was not pulled from Docker Hub, display an error message and exit with non-zero exit status
